### ISSUE TITLE
data race from unit test

### ISSUE THREAD

**Issue by kardianos** [denisenkom/go-mssqldb#680]
https://ci.appveyor.com/project/denisenkom/go-mssqldb/builds/39994729/job/nq0golv5qehgjeaq

```
go test -race -cpu 4 -coverprofile=coverage.txt -covermode=atomic
2021/07/14 19:29:23 got token tokenLoginAck
2021/07/14 19:29:23 got token tokenEnvChange
2021/07/14 19:29:23 got token tokenDone
2021/07/14 19:29:23 got DONE or DONEPROC status=0
2021/07/14 19:29:23 
SET XACT_ABORT ON; -- 16384
SET ANSI_NULLS ON; -- 32
SET ARITHIGNORE ON; -- 128
2021/07/14 19:29:23 got token tokenEnvChange
2021/07/14 19:29:23 got token tokenDone
2021/07/14 19:29:23 got DONE or DONEPROC status=1
2021/07/14 19:29:23 got token tokenDone
2021/07/14 19:29:23 got DONE or DONEPROC status=1
2021/07/14 19:29:23 got token tokenDone
2021/07/14 19:29:23 got DONE or DONEPROC status=0
2021/07/14 19:29:23 
select Options = @@OPTIONS;
2021/07/14 19:29:23 got token tokenColMetadata
2021/07/14 19:29:23 got token tokenRow
2021/07/14 19:29:23 got token tokenDone
2021/07/14 19:29:23 got DONE or DONEPROC status=16
2021/07/14 19:29:23 (1 row(s) affected)
2021/07/14 19:29:23 
SET XACT_ABORT ON; -- 16384
SET ANSI_NULLS ON; -- 32
SET ARITHIGNORE ON; -- 128
2021/07/14 19:29:23 got token tokenEnvChange
2021/07/14 19:29:23 got token tokenDone
2021/07/14 19:29:23 got DONE or DONEPROC status=1
2021/07/14 19:29:23 got token tokenDone
2021/07/14 19:29:23 got DONE or DONEPROC status=1
2021/07/14 19:29:23 got token tokenDone
2021/07/14 19:29:23 got DONE or DONEPROC status=0
==================
WARNING: DATA RACE
Read at 0x00c0000dc143 by goroutine 56:
  testing.(*common).logDepth()
      c:/go113/src/testing/testing.go:665 +0xa8
  testing.(*common).Logf()
      c:/go113/src/testing/testing.go:658 +0x96
  testing.(*T).Logf()
      <autogenerated>:1 +0x7c
  github.com/denisenkom/go-mssqldb.(*testLogger).Printf()
      C:/gopath/src/github.com/denisenkom/go-mssqldb/tds_test.go:253 +0xa2
  github.com/denisenkom/go-mssqldb.optionalLogger.Printf()
      C:/gopath/src/github.com/denisenkom/go-mssqldb/log.go:18 +0xb5
  github.com/denisenkom/go-mssqldb.processSingleResponse()
      C:/gopath/src/github.com/denisenkom/go-mssqldb/token.go:655 +0x2d5
Previous write at 0x00c0000dc143 by goroutine 45:
  testing.tRunner.func1()
      c:/go113/src/testing/testing.go:900 +0x35a
  testing.tRunner()
      c:/go113/src/testing/testing.go:913 +0x1c2
Goroutine 56 (running) created at:
  github.com/denisenkom/go-mssqldb.startReading()
      C:/gopath/src/github.com/denisenkom/go-mssqldb/token.go:772 +0xab
  github.com/denisenkom/go-mssqldb.(*Stmt).processExec()
      C:/gopath/src/github.com/denisenkom/go-mssqldb/mssql.go:695 +0x105
  github.com/denisenkom/go-mssqldb.(*Stmt).exec()
      C:/gopath/src/github.com/denisenkom/go-mssqldb/mssql.go:688 +0x125
  github.com/denisenkom/go-mssqldb.(*Stmt).ExecContext()
      C:/gopath/src/github.com/denisenkom/go-mssqldb/mssql.go:1022 +0x305
  database/sql.ctxDriverStmtExec()
      c:/go113/src/database/sql/ctxutil.go:65 +0x2c7
  database/sql.resultFromStatement()
      c:/go113/src/database/sql/sql.go:2435 +0x1cd
  database/sql.(*DB).execDC()
      c:/go113/src/database/sql/sql.go:1538 +0x33d
  database/sql.(*DB).exec()
      c:/go113/src/database/sql/sql.go:1499 +0x17b
  database/sql.(*DB).ExecContext()
      c:/go113/src/database/sql/sql.go:1477 +0xf5
  github.com/denisenkom/go-mssqldb.TestDisconnect1()
      C:/gopath/src/github.com/denisenkom/go-mssqldb/queries_test.go:2193 +0x60c
  testing.tRunner()
      c:/go113/src/testing/testing.go:909 +0x1a0
Goroutine 45 (running) created at:
  testing.(*T).Run()
      c:/go113/src/testing/testing.go:960 +0x658
  testing.runTests.func1()
      c:/go113/src/testing/testing.go:1202 +0xad
  testing.tRunner()
      c:/go113/src/testing/testing.go:909 +0x1a0
  testing.runTests()
      c:/go113/src/testing/testing.go:1200 +0x528
  testing.(*M).Run()
      c:/go113/src/testing/testing.go:1117 +0x306
  main.main()
      _testmain.go:410 +0x33e
==================
2021/07/14 19:29:34 [{Hello} {World} {TVP}]
FAIL
coverage: 77.2% of statements
exit status 1
FAIL	github.com/denisenkom/go-mssqldb	12.365s
```

**Comment by kardianos** [comment]
I think this is due to the testing code, in the `testLogger` that logs to a test on multiple goroutines, which is a not good I think. One way to fix this would be to remove goroutine use from reading results.

**Comment by shueybubbles** [comment]
Is each test function guaranteed to be running in a process where no other tests functions are competing for access to the driver instance at the same time and could be assigning different t instances to the instance-wide logger? I don't know exactly how go test manages concurrency at that level. Conceptually it seems bad to assign test-scoped object to a globally scoped logger, and just avoiding goroutine usage on read in the driver is a mitigation but not a fix. The tests need some other way to read connection logs that are specific to their connection instance.

**Comment by shueybubbles** [comment]
i think we can live without a separate closer function as it's likely tests will fail if the dev doesn't include a stopLogging call and we'll try to be good code reviewers.

One little thing - while you are touching the logging functions in the tests, add a t.Helper() call in them so the test traces don't use their line numbers as the trace origin.

**Comment by fineol** [comment]
I included the changes to resolve this issue in PR #691, because it contains other logging-related updates.

I ended up not including the t.Helper() call for two reasons:
1. Go 1.8 does not support t.Helper(), and this driver must build under Go 1.8. I could have separated out the testLogger implementation into its own set of files and applied appropriate build tags to get around this problem, however the next point argues against the effort.
2. The majority of the calls to testLogger.Printf/Println are not made directly from test functions but indirectly through the driver's optionalLogger Logger interface. Because of this, adding t.Helper() calls merely changes those log entries from tds_test.go:274 to log:70. That doesn't help much.

To compensate for the above deficiencies, I added a new field to the log message. This field contains the file and line number of the actual caller in the same form as the t.Log adds automatically, but appearing in square brackets after the timestamp. In the following example "tds_test.go:274" is added automatically by t.Log (there is nothing we can do about that) while [token.go:659] is the new field I included that gives the actual caller location:

`tds_test.go:274: 2021-09-19 17:32:20.684711888 -0400 EDT m=+0.388791148 [token.go:659]: ERROR: BeginRead failed EOF`
